changelog/0058-upgrade-to-python-3-10.yml

```diff
@@ -0,0 +1,34 @@
+title: Upgrade to Python 3.10
+key: upgrade-to-python-3-10
+date: 2022-10-19
+optional_per_env: no
+min_commcare_version:
+max_commcare_version:
+context: |
+  Follow these steps to install and use Python 3.10 on your control machine(s) by December 5th, 2022.
+
+details: |
+  As part of our ongoing effort to keep CommCare HQ up to date with the latest tools and
+  libraries we are requiring that commcare-cloud run on Python 3.10. Python 3.6 has reached end of life,
+  and we are requiring that everyone move to Python 3.10 by December 5th, 2022.
+
+update_steps: |
+  Follow the steps below to properly install Python 3.10 and its dependencies.
+
+  $ sudo add-apt-repository -y ppa:deadsnakes/ppa
+  $ sudo apt update
+  $ sudo apt-get -y install python3.10 python3.10-dev python3.10-distutils python3.10-venv libffi-dev
+
+  Once Python 3.10 is installed on your control machine, run:
+  
+  $ update-code
+
+  which will pull the latest version of commcare-cloud and trigger the creation and activation of a Python 3.10 virtual environment.
+
+  Confirm the active virtual environment is using Python 3.10
+
+  $ python --version
```

Comment:
I'd like us to start using and referencing `python3`, but I'm open to discuss this.

Reply:
I think it would be best practice based on https://peps.python.org/pep-0394/ and something we should get in the habit of doing (same as `python3 -m pip` instead of `pip`), since we moved away from 2.7 and only support 3,

Reply:
That PEP pertains to naming of the system python on Unixes. In a context such as this one where a virtualenv is expected to be activated, using `python` is preferred. From the PEP:

- While far from being universally available, `python` remains the preferred spelling for explicitly invoking Python, as this is the spelling that virtual environments make consistently available across different platforms and Python installations.
- When a virtual environment (...) is active, the `python` command should refer to the virtual environment’s interpreter and should always be available.
- As of 2019, activating a Python virtual environment (or its functional equivalent) prior to script execution is one way to obtain a consistent cross-platform and cross-distribution experience.
- If these conventions are adhered to, it will become the case that the `python` command is only executed in an interactive manner as a user convenience, or else when using a virtual environment or similar mechanism.

Reply:
I see, I missed these updates. I see in "History of this PEP" that the rationale has changed in the presence of virtualenvs and also following the deprecation of Python 2.
I'm still uncertain about the "hybrid" state of our documentation and scripts, as some rely on the system version which, being from 18.04 with 2.7 still installed (yes, I know :-) ), creates some inconsistencies (but that would be true also for newer systems, where the presence of `python` is not guaranteed and its behavior undefined).
Also, I think the recommendation for the shebang (i.e. set it explicitly to `python3`) remains, and I'd probably prefer us to be consistent in this usage/nomenclature, as we're not dealing only with end users (as defined in the PEP) but also with a more complex system/setup.

Reply:
Sure, for non-virtualenv dependent scripts the best practice is to use `/usr/bin/python3`. In a virtualenv context, the PEP is explicit:
> For scripts that are only expected to be run in an activated virtual environment, shebang lines can be written as `#!/usr/bin/env python`, as this instructs the script to respect the active virtual environment.

Reply:
Ok, so we should use/reference `python`/`#!/usr/bin/env python` for everything related to our venv (almost everything) and `python3`/`#!/usr/bin/python3` for the few things we have outside of it?